File: .tmp_sc4013/java_22/3947_maximum_number_of_items_from_sale_ii.java
// CONFIG class=Solution method=maxItems types=None
// LeetCode 3947 - Maximum Number of Items From Sale II
// https://leetcode.com/problems/maximum-number-of-items-from-sale-ii/

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

class Solution {
    public int maxItems(int[][] items, int budget) {
        int n = items.length;
        int[] frequency = new int[n + 1];
        int minimumPrice = items[0][1];
        for (int[] item : items) {
            frequency[item[0]]++;
            minimumPrice = Math.min(minimumPrice, item[1]);
        }
        List<int[]> batches = new ArrayList<>();
        for (int[] item : items) {
            int gain = 0;
            for (int multiple = item[0]; multiple <= n; multiple += item[0]) gain += frequency[multiple];
            gain--;
            if (gain > 0 && item[1] < 2 * minimumPrice) batches.add(new int[] { item[1], gain });
        }
        batches.sort(Comparator.comparingInt(a -> a[0]));
        long remaining = budget;
        long answer = budget / minimumPrice;
        long boosted = 0;
        for (int[] current : batches) {
            long count = current[1];
            long affordable = remaining / current[0];
            if (affordable < count) count = affordable;
            remaining -= count * current[0];
            boosted += count;
            long total = 2 * boosted + remaining / minimumPrice;
            if (total > answer) answer = total;
            if (count < current[1]) break;
        }
        return (int) answer;
    }
}

File: .tmp_py4013/batch_18_dump/3947_maximum_number_of_items_from_sale_ii.js
// LeetCode 3947 - Maximum Number of Items From Sale II
// https://leetcode.com/problems/maximum-number-of-items-from-sale-ii/

var maxItems = function(items, budget) {
    const n = items.length;
    const frequency = new Array(n + 1).fill(0);
    let minimumPrice = items[0][1];
    for (const item of items) {
        frequency[item[0]]++;
        minimumPrice = Math.min(minimumPrice, item[1]);
    }
    const batches = [];
    for (const item of items) {
        let gain = 0;
        for (let multiple = item[0]; multiple <= n; multiple += item[0]) gain += frequency[multiple];
        gain--;
        if (gain > 0 && item[1] < 2 * minimumPrice) batches.push([item[1], gain]);
    }
    batches.sort((a, b) => a[0] - b[0]);
    let remaining = budget;
    let answer = Math.floor(budget / minimumPrice);
    let boosted = 0;
    for (const current of batches) {
        let count = current[1];
        const affordable = Math.floor(remaining / current[0]);
        if (affordable < count) count = affordable;
        remaining -= count * current[0];
        boosted += count;
        const total = 2 * boosted + Math.floor(remaining / minimumPrice);
        if (total > answer) answer = total;
        if (count < current[1]) break;
    }
    return answer;
};
